src/tty.rs

```diff
@@ -202,17 +203,66 @@ fn get_pw_entry(buf: &mut [i8; 1024]) -> Passwd {
     }
 }
 
+fn parse_shell(raw: Option<&String>) -> Option<Vec<CString>> {
+    match raw {
+        Some(raw) => {
+            let mut argv = Vec::new();
+            let mut st = 0; // Single tick '
+            let mut dt = 0; // Double tick "
+            let mut current_arg = String::new();
+            for c in raw.chars() {
+                match c {
+                    '\'' => st = (st + 1) % 2,
+                    '"' => dt = (dt + 1) % 2,
+                    ' ' | '\t' => {
+                        if st == 0 && dt == 0 {
+                            if let Ok(arg) = CString::new(current_arg.as_bytes()) {
+                                argv.push(arg);
+                            }
+                            current_arg = String::new();
+                        } else {
+                            current_arg.push(c);
+                        }
+                    },
+                    _ => current_arg.push(c),
+                }
+            }
+            if let Ok(arg) = CString::new(current_arg.as_bytes()) {
+                argv.push(arg);
+            }
+            Some(argv)
+        },
+        None => None
+    }
+}
+
 /// Exec a shell
-fn execsh(config: &Config) -> ! {
+fn execsh(config: &Config, options: &Options) -> ! {
     let mut buf = [0; 1024];
     let pw = get_pw_entry(&mut buf);
 
-    let shell = match config.shell() {
-        Some(shell) => match shell.to_str() {
-            Some(shell) => shell,
-            None => die!("Invalid shell value")
+    let argv = match options.shell() {
+        Some(argv) => Some(argv),
+        None => config.shell(),
+    };
+    let argv = parse_shell(argv);
+    let (shell, argv) = match argv {
+        Some(ref argv) => {
+            let mut res = Vec::new();
+            let shell = match argv.first() {
```

Comment:
Yeah, I'm not really proud of that logic. We want the shell to be set in three places right?

1. command line
2. configuration file
3. passwd

But I will try and rewrite the code so that it is more obvious.